cdap-app-templates/cdap-etl/cdap-etl-batch/src/main/java/io/cdap/cdap/etl/batch/StructuredRecordComparator.java

```diff
@@ -0,0 +1,161 @@
+ *
+ * Records are first compared by their schemas. If their schemas are the same, they are compared field by field.
+ *
+ * Record fields are compared first by their schema fields. If the schema fields are the same, they are compared by
+ * value.
+ *
+ * Numbers, boolean, and strings are compared with their normal java compare methods. A null value is less than
+ * a non-null value.
+ *
+ * Arrays are compared by their values in order.
+ *
+ * Bytes are compared lexicographically.
+ */
+public class StructuredRecordComparator implements Comparator<StructuredRecord> {
+  private static final Comparator<Schema> SCHEMA_COMPARATOR = new SchemaComparator();
+
+  @Override
+  public int compare(StructuredRecord r1, StructuredRecord r2) {
+    if (r1 == null && r2 != null) {
+      return -1;
+    } else if (r1 != null && r2 == null) {
+      return 1;
+    } else if (r1 == null) {
+      return 0;
+    }
+
+    int comp = SCHEMA_COMPARATOR.compare(r1.getSchema(), r2.getSchema());
+    if (comp != 0) {
+      return comp;
+    }
+
+    // both records must have the same fields, otherwise their schemas would be different
+    //noinspection ConstantConditions
+    for (Schema.Field field : r1.getSchema().getFields()) {
+      comp = compareFields(field.getName(), field.getSchema(), r1.get(field.getName()), r2.get(field.getName()));
+      if (comp != 0) {
+        return comp;
+      }
+    }
+
+    return 0;
+  }
+
+  private int compareFields(String fieldName, Schema schema, Object val1, Object val2) {
+    if (val1 == null && val2 != null) {
+      return -1;
+    } else if (val1 != null && val2 == null) {
+      return 1;
+    } else if (val1 == null) {
+      return 0;
+    }
+
+    switch (schema.getType()) {
+      case NULL:
+        return 0;
+      case INT:
+        return Integer.compare((int) val1, (int) val2);
+      case LONG:
+        return Long.compare((long) val1, (long) val2);
+      case FLOAT:
+        return Float.compare((float) val1, (float) val2);
+      case DOUBLE:
+        return Double.compare((double) val1, (double) val2);
+      case STRING:
+        return ((String) val1).compareTo((String) val2);
+      case BOOLEAN:
+        return Boolean.compare((boolean) val1, (boolean) val2);
+      case ENUM:
+        return Integer.compare(((Enum) val1).ordinal(), ((Enum) val2).ordinal());
+      case BYTES:
+        return Bytes.compareTo(asBytes(fieldName, val1), asBytes(fieldName, val2));
+      case RECORD:
+        return compare((StructuredRecord) val1, (StructuredRecord) val2);
+      case ARRAY:
+        return compareArrays(fieldName, schema.getComponentSchema(), val1, val2);
+      case UNION:
+        throw new IllegalArgumentException(String.format("Field '%s' is a union. Cannot compare union fields",
+                                                         fieldName));
+      case MAP:
+        throw new IllegalArgumentException(String.format("Field '%s' is a union. Cannot compare union fields",
```

Comment:
union/map

Reply:
why can't maps be compared?

Reply:
It would involve sorting all the map contents, though I suppose this extra cost only happens if somebody explicitly decides to use a map as part of a group key.

I'll add it, though it will also involve refactoring out the logic into a comparator per schema type in order to sort the map keys.

Reply:
Not necessarily involved sorting. You can 
1) compare the size of the maps
2) iterate over the entries of the first map and look up each one in the second map. 

I agree it is an uncommon case. I guess supporting union is more important.

Reply:
I guess the error message would also matter here. When a pipeline designed creates a group-by with such a group key schema, they should get an error right away. Whereas this error would be only thrown at runtime when the data is compared.

Reply:
Without sorting, we can have maps where m1 < m2 and m2 < m1, and also situations where m1 < m2 but then a later comparison returns the reverse.
In any case, I think we can still support it in the app, and specific plugins can decide whether or not to allow it or not.

Reply:
I guess you are right, at least one keyset must be sorted.